service/consul/datacenterWatch.go

```diff
@@ -1,109 +1,229 @@
-						keys[key] = true
-
-						// don't create new instancer if it is already saved in environment's instancers
-						if currentInstancers.Has(key) {
-							continue
-						}
-
-						// don't create new instancer if it was already created and added to the new instancers map
-						if instancersToAdd.Has(key) {
-							logger.Log(level.Key(), level.WarnValue(), logging.MessageKey(), "skipping duplicate watch", "service", w.Service, "tags", w.Tags, "passingOnly", w.PassingOnly, "datacenter", w.QueryOptions.Datacenter)
-							continue
-						}
-
-						// create new instancer and add it to the map of instancers to add
-						instancersToAdd.Set(key, newInstancer(logger, client, w))
-					}
+			d.UpdateInstancers(datacenters)
+
+		}
+
+	}
+}
+
+func (d *DatacenterWatcher) UpdateInstancers(datacenters []string) {
+	keys := make(map[string]bool)
+	instancersToAdd := make(service.Instancers)
+
+	options := d.options
+	environment := d.environment
+	logger := d.logger
+	currentInstancers := environment.Instancers()
+
+	for _, w := range options.watches() {
+		if w.CrossDatacenter {
+			for _, datacenter := range datacenters {
+
+				//check if datacenter is part of inactive datacenters list
+				d.lock.RLock()
+				_, found := d.inactiveDatacenters[datacenter]
+				d.lock.RUnlock()
+
+				if found {
+					logger.Log(level.Key(), level.InfoValue(), logging.MessageKey(), "datacenter set as inactive", "datacenter name: ", datacenter)
+					continue
 				}
+				w.QueryOptions.Datacenter = datacenter
+
+				// create keys for all datacenters + watched services
+				key := newInstancerKey(w)
+				keys[key] = true
+
+				// don't create new instancer if it is already saved in environment's instancers
+				if currentInstancers.Has(key) {
+					continue
+				}
+
+				// don't create new instancer if it was already created and added to the new instancers map
+				if instancersToAdd.Has(key) {
+					logger.Log(level.Key(), level.WarnValue(), logging.MessageKey(), "skipping duplicate watch", "service", w.Service, "tags", w.Tags, "passingOnly", w.PassingOnly, "datacenter", w.QueryOptions.Datacenter)
+					continue
+				}
+
+				// create new instancer and add it to the map of instancers to add
+				instancersToAdd.Set(key, newInstancer(logger, d.environment.Client(), w))
 			}
+		}
+	}
+
+	environment.UpdateInstancers(keys, instancersToAdd)
+
+}
 
-			environment.UpdateInstancers(keys, instancersToAdd)
+func (d *DatacenterWatcher) DatacentersListener() chrysom.ListenerFunc {
+	return func(items []model.Item) {
+		updateInactiveDatacenters(items, d.inactiveDatacenters, &d.lock)
 
+		datacenters, err := getDatacenters(d.logger, d.environment.Client(), d.options)
+
+		if err == nil {
+			d.UpdateInstancers(datacenters)
 		}
 
 	}
 }
+
+func updateInactiveDatacenters(items []model.Item, inactiveDatacenters map[string]bool, lock *sync.RWMutex) {
+	chrysomMap := make(map[string]bool)
+	for _, item := range items {
+		datacenterName := item.Data["name"].(string)
```

Comment:
a struct should be created so we can leverage json.Marshal to populate the struct. 
Such as:
```golang
type DatacenterFilter struct {
    Name string
    Inactive bool
}
```

This allows for flexibility for updating the filtering for later.